mqtt/mqtt-bridge/src/upstream/rpc/local.rs

```diff
@@ -0,0 +1,203 @@
+use async_trait::async_trait;
+use bson::{doc, Document};
+use bytes::buf::BufExt;
+use lazy_static::lazy_static;
+use mqtt3::Event;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    client::{EventHandler, Handled},
+    pump::{PumpHandle, PumpMessage},
+    upstream::{CommandId, RemoteUpstreamPumpEvent, RpcCommand},
+};
+
+use super::RpcError;
+
+/// An RPC handlers that reposible to connect to part of the bridge which
+/// connects to local broker. It receives RPC commands on a special topic,
+/// converts it to a `RpcCommand` and sends to remote pump as a `PumpMessage`.
+pub struct LocalRpcHandler {
+    remote_pump: PumpHandle<RemoteUpstreamPumpEvent>,
+}
+
+impl LocalRpcHandler {
+    pub fn new(remote_pump: PumpHandle<RemoteUpstreamPumpEvent>) -> Self {
+        Self { remote_pump }
+    }
+}
+
+#[async_trait]
+impl EventHandler for LocalRpcHandler {
+    type Error = RpcError;
+
+    async fn handle(&mut self, event: &Event) -> Result<Handled, Self::Error> {
+        if let Event::Publication(publication) = event {
+            if let Some(command_id) = capture_command_id(&publication.topic_name) {
+                let doc = Document::from_reader(&mut publication.payload.clone().reader())?;
+                match bson::from_document(doc)? {
+                    VersionedRpcCommand::V1(command) => {
+                        let event =
+                            RemoteUpstreamPumpEvent::RpcCommand(command_id.clone(), command);
+                        let msg = PumpMessage::Event(event);
+                        self.remote_pump
+                            .send(msg)
+                            .await
+                            .map_err(|e| RpcError::SendToRemotePump(command_id, e))?;
+
+                        return Ok(Handled::Fully);
+                    }
+                }
+            }
+        }
+
+        Ok(Handled::Skipped)
+    }
+}
+
+fn capture_command_id(topic_name: &str) -> Option<CommandId> {
+    lazy_static! {
+        static ref RPC_TOPIC_PATTERN: Regex = Regex::new("\\$edgehub/rpc/(?P<command_id>[^/ ]+)$")
+            .expect("failed to create new Regex from pattern");
+    }
+
+    RPC_TOPIC_PATTERN
+        .captures(topic_name)
+        .and_then(|captures| captures.name("command_id"))
+        .map(|command_id| command_id.as_str().into())
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "version")]
+enum VersionedRpcCommand {
+    V1(RpcCommand),
+}
+
+#[cfg(test)]
+mod tests {
+    use bson::{bson, spec::BinarySubtype};
+    use bytes::Bytes;
+    use matches::assert_matches;
+    use mqtt3::{proto::QoS, ReceivedPublication};
+    use test_case::test_case;
+
+    use super::*;
+
+    #[test]
+    fn it_deserializes_from_bson() {
```

Comment:
If you feel like adding it, maybe a test for parsing a publication.

Reply:
I believe `it_handles_rpc_commands` tests is all about parsing publication and making a pump message